### PR TITLE
[flutter_tools] Add support for vmServiceFileInfo when attaching

### LINKED ISSUE #4577: vmServiceInfoFile does not work with Flutter SDK debug adapter
Content of the vmService file:

```
{
  "uri": "http://127.0.0.1:59168/t1E27KPVN5w=/"
}%      
```

[logs-new-adapter (not working).txt](https://github.com/Dart-Code/Dart-Code/files/11681307/logs-new-adapter.not.working.txt)

[logs-old-adapter (working).txt](https://github.com/Dart-Code/Dart-Code/files/11681308/Dart-Code-Log-2023-05-03.21-34-07.txt)

### PR DESCRIPTION
**Pull request by DanTup**
When building the new SDK DAPs, this functionality was missed from the Flutter adapter (but added to the Dart CLI adapter).

As well as passing a VM Service URI directly, we support passing a file that can be polled for it.

This uses the same mechanism we use to obtain the VM Service URI from a Dart debug session (we run `dart --write-service-info=foo.json my_file.dart` and then poll that file which the VM will write) and is useful for users that have their own mechanism for launching an app (for example using custom Flutter embedders - see https://github.com/Dart-Code/Dart-Code/issues/3353) to provide a VM Service URI once the app is up and running.

Fixes https://github.com/Dart-Code/Dart-Code/issues/4577.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/debug_adapters/README.md [hunk 1]
```diff
@@ -40,7 +40,10 @@ Arguments specific to `launchRequest` are:
 
 Arguments specific to `attachRequest` are:
 
-- `String? vmServiceUri` - the VM Service URI to attach to (if not supplied, Flutter will try to discover it from the device)
+- `String vmServiceInfoFile` - the file to read the VM Service info from \*
+- `String vmServiceUri` - the VM Service URI to attach to \*
```
[the remark this answers is not in the document]
Comment (DanTup) [reply packages/flutter_tools/lib/src/debug_adapters/README.md:44]: Oops, no - apparently I pasted the Dart version over this and didn't check it properly (it's wrong there, and I also lost the additional explanatory text). Fixed!

File: packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart [hunk 2]
```diff
@@ -120,14 +121,28 @@ class FlutterDebugAdapter extends FlutterBaseDebugAdapter {
   @override
   Future<void> attachImpl() async {
     final FlutterAttachRequestArguments args = this.args as FlutterAttachRequestArguments;
+    String? vmServiceUri = args.vmServiceUri;
+    final String? vmServiceInfoFile = args.vmServiceInfoFile;
+
+    if (vmServiceUri != null && vmServiceInfoFile != null) {
```
[the remark this answers is not in the document]
Comment (DanTup) [reply packages/flutter_tools/lib/src/debug_adapters/flutter_adapter.dart:127]: Yep, I've made the error text a bit clearer (and also reintroduced this explanation in the readme noted above).

**Comment by DanTup** [comment]
Out of interest - does the autosubmit bot handle review status? Like if I add the tag now, would it just blindly merge when the build is green, or wait for you to approve? 🤔

**Comment by christopherfujino** [comment]
> Out of interest - does the autosubmit bot handle review status? Like if I add the tag now, would it just blindly merge when the build is green, or wait for you to approve? thinking

Unfortunately, I think it would remove the label since it's not approved.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by DanTup** [comment]
> Unfortunately, I think it would remove the label since it's not approved.

Ah, yes. I wonder if it's worth a feature request to just delay until it's reviewed? I think that's what Gerrit does - you can add AutoSubmit to mean "when this is approved, land it without me needing to come back and do it"

> added the label for you 

Thanks! :)

**Comment by christopherfujino** [comment]
> > Unfortunately, I think it would remove the label since it's not approved.
> 
> Ah, yes. I wonder if it's worth a feature request to just delay until it's reviewed? I think that's what Gerrit does - you can add AutoSubmit to mean "when this is approved, land it without me needing to come back and do it"

Good idea ;) https://github.com/flutter/flutter/issues/113638

**Comment by DanTup** [comment]
> Good idea ;) https://github.com/flutter/flutter/issues/113638
> 
> [christopherfujino](https://github.com/christopherfujino) commented [on Oct 18, 2022](https://github.com/flutter/flutter/issues/113638#issue-1413759925)

LOL! Always one step ahead 😁

**Pull request merged**